ibis/common/grounds.py

```diff
@@ -245,3 +231,31 @@ def __cached_equals__(self, other):
             self.__cache__[key] = result
 
         return result
+
+
+# TODO(kszucs): need to find a nice name for it
+class Concrete(Comparable, Annotable, Immutable):
+
+    __slots__ = ("__args__", "__precomputed_hash__")
+
+    def __post_init__(self):
+        # optimizations to store frequently accessed generic properties
+        arguments = tuple(getattr(self, name) for name in self.__argnames__)
+        hashvalue = hash((self.__class__, arguments))
+        object.__setattr__(self, "__args__", arguments)
+        object.__setattr__(self, "__precomputed_hash__", hashvalue)
+        super().__post_init__()
+
+    def __hash__(self):
+        return self.__precomputed_hash__
+
+    def __equals__(self, other):
+        return self.__args__ == other.__args__
+
+    @property
+    def args(self):
+        return self.__args__
+
+    @property
+    def argnames(self):
```

Comment:
Can we have a property that zips `argnames` and `args` together? This is a common enough pattern that we should make it convenient IMO.

Reply:
I added an `as_dict()` method to `Annotable`: https://github.com/ibis-project/ibis/pull/4493/files#diff-1ba248744f86eb25b4e863648c19353294f34c47b637d1fe88d66eccfc1e1119R152

Perhaps it would be better just to return with the items view: `zip(self.__argnames__, self.__args__)`

Reply:
Yeah, it's easy to just call `dict` IMO

Reply:
I'm okay with `as_dict` for now.